### PR TITLE
Converting tests to junit5

### PR DESCRIPTION
**Pull request by StefanSpieker**
This is mainly for discussion. I started to convert some easy classes to junit 5. I will convert some classes where it is more complicated.

### Proposed changelog entries

* Developer: Added possibility to add JUnit5 tests

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — COMMENTED** [review]
Seems fine to me? are you planning to do more on this? (i.e. it's in draft)

**Comment by StefanSpieker** [comment]
I wanted to add also a conversion from a parameterized class. I would like to continue on this. But I think the PR might get too big.

**Review by timja — APPROVED** [review]

File: core/pom.xml
Comment (MarkEWaite) [inline core/pom.xml:46]: Any particular reason you've chosen 5.5.2 instead of the 5.6.2 release from April 2020?
```suggestion
    <junit.jupiter.version>5.6.2</junit.jupiter.version>
```

Comment (StefanSpieker) [reply core/pom.xml]: I just copied the version from the cli. I plan to move the version to the parent-pom or the bom.

**Review by rsandell — COMMENTED** [review]
Generally I am against changing/updating tests for changing's sake. If the tests starts to fail you can get confused if it is the change in library or an actual regression.
I have no issues on introducing a new testing library for new tests though :)

File: core/pom.xml
Comment (alecharp) [inline core/pom.xml:485]: If we want to migrate our test base, we should remove this library all together.

**Review by alecharp — CHANGES_REQUESTED** [review]
Junit 4 library should be removed. There is a module in Junit 5 which provide the required migration tools to get ride of this library.

File: core/pom.xml
Comment (StefanSpieker) [reply core/pom.xml]: Thanks for pointing that out, I removed junit4 and added the vintage engine

**Comment by StefanSpieker** [comment]
fixed merge conflict, please remove the label.

File: core/pom.xml
Comment (jglick) [inline core/pom.xml:43]: I take it you have not tried to convert the `test` module with `JenkinsRule`? That is where most test coverage lies, but I wonder if a migration is practical.

TBH other than `assertThrows` (which could equally be a static utility method anywhere and used with JUnit 4) I have not seen a compelling advantage to using JUnit 5. It does not correct the main sin of JUnit 4, which is the abuse of annotations.

Comment (timja) [reply core/pom.xml:43]: `assertThrows` is in junit 4.13

Comment (StefanSpieker) [reply core/pom.xml:43]: Actually I took a look at the ``test`` module, but this might take a little longer, because the JenkinsRule has to be changed to an Extension. If this would be needed before this PR can be merged, I can spend there some time.

But to answer the question about JUnit5 in general: I think there are a few things which are easier with JUnit5 than JUnit4, but that is only my opinion. Some examples:

- [Tests enabled for a specified JRE range ](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/condition/EnabledForJreRange.html)
- dynamic tests with @TestFactory

Another nice blog I stumbled upon is this one: [why you should migrate to junit-5](https://medium.com/riag-digital-techblog/why-you-should-migrate-to-junit-5-15438f5c56f)

Comment (jglick) [reply core/pom.xml:43]: > If this would be needed before this PR can be merged

No, just wondering if that is part of the plan.

`@TestFactory` does sound easier to work with than JUnit 4 parameterized tests.

My experience with tests in @jenkinsci is that it is tough to make various systems interoperate cleanly, due to the misguided attempt by test frameworks to make things look declarative using annotations and reflection. Thus there are problems mixing `JenkinsRule` or `RestartableJenkinsRule` with timeouts, abstract base classes, `LoggerRule`, `docker-fixtures`, before/after class/method, parameters, ad nauseam. As a developer I know exactly what I want to happen in what order but the framework, or perhaps misuse of the framework, is an obstacle rather than a convenience.

**Review by oleg-nenashev — APPROVED** [review]
I do not have a strong opinion about migratingthe Jenkins Test Harness codebase to JUnit 5. Indeed it looks like a big effort with limited benefits. SAnd definitely there is no reason to migrate Docker Fixtures, we should rather go testcontainers if we do such migration.

For unit tests there is no practical difference, I am +1 for moving to JUnit 5 there

File: core/pom.xml
Comment (oleg-nenashev) [reply core/pom.xml:43]: @jglick would you approve it after the discussion?

Comment (jglick) [reply core/pom.xml:43]: I was never blocking anything here, just discussing possibilities and plans.

**Comment by oleg-nenashev** [comment]
@StefanSpieker Hi. Could you please resolve the merge conflict? I think we can land it after that

File: core/src/test/java/hudson/PluginWrapperTest.java
Comment (stellargo) [inline core/src/test/java/hudson/PluginWrapperTest.java:13]: Importing necessary classes is preferred

File: core/src/test/java/hudson/PluginWrapperTest.java
Comment (stellargo) [inline core/src/test/java/hudson/PluginWrapperTest.java:17]: I am not sure if this is used

Comment (StefanSpieker) [reply core/src/test/java/hudson/PluginWrapperTest.java:19]: It is, because of Line 198:    @Issue("JENKINS-52665")

File: core/src/test/java/hudson/PluginWrapperTest.java
Comment (StefanSpieker) [reply core/src/test/java/hudson/PluginWrapperTest.java]: Thanks, removed the star import

**Comment by StefanSpieker** [comment]
@oleg-nenashev : Resolved merge conflict

File: core/src/test/java/hudson/PluginWrapperTest.java
Comment (stellargo) [reply core/src/test/java/hudson/PluginWrapperTest.java:19]: Ahhh, nevermind then :)

**Comment by timja** [comment]
This PR is now ready for merge, we'll merge it after 24 hours if there's no negative feedback

**Pull request merged**
